intermittent failure with npm install running lockdown on jenkins box with cb() never called!

on jenkins run this:
npm install
cd automation-tests
npm install
mkdir results
./scripts/run-all.js --platform win7_firefox_16 --parallel 10 --output xunit --env dev

results:
    For the 63 builds (mac, win linux) i got this 5 times.
    Edwin@ewong-16235:~/github/edmoz/jenkins-cli$ grep -B 3 -irn 'never called' logs/identity.browserid.\* 
    logs/identity.browserid.dev.linux_firefox_16.2089.txt-198-npm http 200 http://127.0.0.1:51174/htmlparser/-/htmlparser-1.7.6.tgz
    logs/identity.browserid.dev.linux_firefox_16.2089.txt-199-npm http 200 http://127.0.0.1:51174/connect-cookie-session/-/connect-cookie-session-0.0.2.tgz
    logs/identity.browserid.dev.linux_firefox_16.2089.txt-200-npm http 200 http://127.0.0.1:51174/connect-logger-statsd/-/connect-logger-statsd-0.0.1.tgz
    logs/identity.browserid.dev.linux_firefox_16.2089.txt:201:npm ERR! cb() never called!
    --
    logs/identity.browserid.dev.win7_firefox_16.2842.txt-198-npm http 200 http://127.0.0.1:45963/which/-/which-1.0.5.tgz
    logs/identity.browserid.dev.win7_firefox_16.2842.txt-199-npm http 200 http://127.0.0.1:45963/htmlparser/-/htmlparser-1.7.6.tgz
    logs/identity.browserid.dev.win7_firefox_16.2842.txt-200-npm http 200 http://127.0.0.1:45963/jshint/-/jshint-0.9.1.tgz
    logs/identity.browserid.dev.win7_firefox_16.2842.txt:201:npm ERR! cb() never called!
    --
    logs/identity.browserid.dev.win7_firefox_16.2851.txt-864-  SOLINK_MODULE(target) Release/obj.target/iconv.node: Finished
    logs/identity.browserid.dev.win7_firefox_16.2851.txt-865-  COPY Release/iconv.node
    logs/identity.browserid.dev.win7_firefox_16.2851.txt-866-make: Leaving directory `/var/lib/jenkins/jobs/identity.browserid.dev.win7_firefox_16/workspace/node_modules/nodemailer/node_modules/mailcomposer/node_modules/mimelib/node_modules/encoding/node_modules/iconv/build'
    logs/identity.browserid.dev.win7_firefox_16.2851.txt:867:npm ERR! cb() never called!
    --
    logs/identity.browserid.dev.win7_firefox_16.2852.txt-198-npm http 200 http://127.0.0.1:56867/jshint/-/jshint-0.9.1.tgz
    logs/identity.browserid.dev.win7_firefox_16.2852.txt-199-npm http 200 http://127.0.0.1:56867/winston/-/winston-0.6.2.tgz
    logs/identity.browserid.dev.win7_firefox_16.2852.txt-200-npm http 200 http://127.0.0.1:56867/htmlparser/-/htmlparser-1.7.6.tgz
    logs/identity.browserid.dev.win7_firefox_16.2852.txt:201:npm ERR! cb() never called!
    --
    logs/identity.browserid.dev.win7_firefox_16.2860.txt-844-  SOLINK_MODULE(target) Release/obj.target/iconv.node: Finished
    logs/identity.browserid.dev.win7_firefox_16.2860.txt-845-  COPY Release/iconv.node
    logs/identity.browserid.dev.win7_firefox_16.2860.txt-846-make: Leaving directory`/var/lib/jenkins/jobs/identity.browserid.dev.win7_firefox_16/workspace/node_modules/nodemailer/node_modules/mailcomposer/node_modules/mimelib/node_modules/encoding/node_modules/iconv/build'
    logs/identity.browserid.dev.win7_firefox_16.2860.txt:847:npm ERR! cb() never called!

Good find. I did some poking around with https://ci.mozilla.org/view/Persona/job/identity.browserid.dev.osx_safari_5/configure to see some things I was interested in (I can remove those later).

Anyways, on the jenkins server, all jobs run as the same user on whatever schedule those jobs run. Since they are the same $HOME then npm will be using the same ~/.npm. On the jenkins box there are browserid jobs and other projects that use node.js. The browserid jobs use `npm-lockdown` which will need to share `/tmp/.npm` as that user if jobs run concurrently.

I set up an awsbox to just loop over the npm install with a clean clone. That works fine for two users running that task at the same time. However, if I interrupt one task in the middle, the other task sometimes reproduces the `npm ERR! cb() never called!`. 

I don't know where the real bug is (npm or lockdown). But we should configure the Jenkins jobs to use isolated per-job locations for the equivalent of $HOME/.npm and $TMPDIR/.npm.

--- a/scripts/lockdown
+++ b/scripts/lockdown
@@ -92,6 +92,13 @@ function rewritePackageMD(json) {
   return JSON.stringify(json);
 }
 
+function copy(from, to) {
+  for (var k in from) {
+    to[k] = from[k];
+  }
+  return to;
+}
+
 var server = http.createServer(function (req, res) {
   if (req.method !== 'GET') {
     return res.end('non GET requests not supported', 501);
@@ -163,15 +170,14 @@ var server = http.createServer(function (req, res) {
 
 server.listen(process.env['LOCKDOWN_PORT'] || 0, '127.0.0.1', function() {
   boundPort = server.address().port;
+  var env = copy(process.env, {
+    NPM_CONFIG_REGISTRY: 'http://127.0.0.1:' + boundPort,
+    NPM_LOCKDOWN_RUNNING: "true"
+  });
 
   var child = exec('npm install', {
-    env: {
-      NPM_CONFIG_REGISTRY: 'http://127.0.0.1:' + boundPort,
-      NPM_LOCKDOWN_RUNNING: "true",
-      PATH: process.env['PATH'],
-      HOME: process.env['HOME']
-    },
-    cwd: process.cwd()
+    cwd: process.cwd(),
+    env: env
   }, function(e) {
     if (warn.length) {
       console.log();

